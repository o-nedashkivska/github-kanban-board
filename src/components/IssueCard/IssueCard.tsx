import { useMemo } from "react";
import { Card, Flex, Typography } from "antd";
import { Draggable } from "react-beautiful-dnd";

import { formatCreatedAt } from "../../utils/formatCreatedAt";

import styles from "./issue-card.module.css";

interface IssueCardProps {
  index: number;
  title: string;
  number: number;
  comments: number;
  assignee: null | Partial<{ login: string }>;
  created_at: Date;
}

const { Text } = Typography;

const IssueCard: React.FC<IssueCardProps> = ({
  title,
  index,
  number,
  comments,
  assignee,
  created_at,
}) => {
  assignee = assignee ?? { login: "No assignee" };

  const activeTime = useMemo(() => formatCreatedAt(created_at), [created_at]);

  return (
    <Draggable draggableId={created_at} index={index}>
      {(provided, snapshot) => (
        <Card
          size="small"
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
          className={styles.card}
          style={{
            ...provided.draggableProps.style,
            backgroundColor: snapshot.isDragging && "#e6f4ff",
          }}
        >
          <Flex vertical>
            <Text>{title}</Text>
            <Text type="secondary">
              #{number} {activeTime}
            </Text>
            <Text type="secondary">
              {assignee.login} | Comments: {comments}
            </Text>
          </Flex>
        </Card>
      )}
    </Draggable>
  );
};

export default IssueCard;
