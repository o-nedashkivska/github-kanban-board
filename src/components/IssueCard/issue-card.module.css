.card {
  width: 90%;
  border: 1px solid black;
  font-weight: 600;
  z-index: 100;
}

.card::before {
  position: absolute;
  top: 1px;
  bottom: 1px;
  left: -4px;
  right: -4px;
  display: block;
  border: 1px solid black;
  border-radius: 8px;
  content: "";
  transform: rotate(0.1deg);
}

.card::after {
  position: absolute;
  top: 1px;
  bottom: 1px;
  left: -4px;
  right: -4px;
  display: block;
  border: 1px solid black;
  border-radius: 8px;
  content: "";
  transform: rotate(-0.7deg);
}
