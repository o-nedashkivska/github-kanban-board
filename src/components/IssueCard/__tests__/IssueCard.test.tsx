import { screen } from "@testing-library/react";

import { renderWithProviders } from "../../../test-utils";

import IssueCard from "../IssueCard";

describe("IssueCard", () => {
  const issues = [
    {
      title: "Title 2",
      index: 1,
      number: 2022,
      comments: 22,
      assignee: { login: "username" },
      created_at: new Date(),
    },
    {
      title: "Title 1",
      index: 0,
      number: 1011,
      comments: 11,
      assignee: null,
      created_at: new Date(),
    },
  ];

  test("displays correct card", () => {
    const cardData = issues[0];
    renderWithProviders(<IssueCard {...cardData} />);

    expect(screen.getByText(cardData.title)).toBeInTheDocument();

    expect(
      screen.getByText(`#${cardData.number} opened today`)
    ).toBeInTheDocument();

    const commentsText = `Comments: ${cardData.comments}`;
    const loginText = cardData.assignee.login;

    expect(
      screen.getByText(`${loginText} | ${commentsText}`)
    ).toBeInTheDocument();
  });

  test("displays correct text when assignee is null", () => {
    const cardData = issues[1];
    renderWithProviders(<IssueCard {...cardData} />);

    const commentsText = `Comments: ${cardData.comments}`;
    expect(
      screen.getByText(`No assignee | ${commentsText}`)
    ).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = renderWithProviders(<IssueCard {...issues[0]} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
