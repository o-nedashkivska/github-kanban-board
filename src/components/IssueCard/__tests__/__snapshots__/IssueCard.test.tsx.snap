// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered ant-card-small card css-dev-only-do-not-override-1kuana8"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-1kuana8 ant-flex-align-stretch ant-flex-vertical"
      >
        <span
          class="ant-typography css-dev-only-do-not-override-1kuana8"
        >
          Title 2
        </span>
        <span
          class="ant-typography ant-typography-secondary css-dev-only-do-not-override-1kuana8"
        >
          #2022 opened today
        </span>
        <span
          class="ant-typography ant-typography-secondary css-dev-only-do-not-override-1kuana8"
        >
          username | Comments: 22
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
